Lookup Table Limit Restore to Recently Modified

Jtang-1 opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->
No UI effects. With this change, incremental syncs will only include lookup table fixtures that have been modified since the last restore time. This will result in the restore being smaller, leading to faster syncs.

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
[USH-5833](https://dimagi.atlassian.net/browse/USH-5833)

Lookup table fixtures don’t change often, and when they don’t, there’s no need to include them in an incremental sync restore.  

This update adds a last_modified field with auto_now=True to `LookupTable`. This ensures `last_modified` is automatically updated with the current UTC timestamp whenever an instance of `LookupTable` is saved or created. One caveat is that [auto_now](https://docs.djangoproject.com/en/5.1/ref/models/fields/#django.db.models.DateField.auto_now) does not update on update() or bulk_update(), as it only triggers on `save()`. However, we also explicitly update the `LookupTable` instances's `last_modified` value whenever the cache is cleared for that table. This occurs whenever a `LookupTableRow` or `LookupTableRowOwner` related to the table is created, updated, or deleted.

### Caching  
As is currently, we cache the serialized XML fixture elements. This includes a set of "global fixtures," which are the same for all users on the domain. Because of this fact, it was prudent to compute the XML for these global fixtures once, store it in blobdb, and serve to users directly from there, skipping the steps of pulling from SQL and serializing as XML. (Credit to Ethan for the well-worded original comment.) The cache is cleared when a lookup table is modified.

With the changes in this PR, if we continue caching all fixtures together, the cache will, in cases where no lookup table is modified, contain either:

1. The most recent list of modified tables, or
2. All tables (if a full sync occurs).
To continue supporting caching, this PR splits it so that each fixture xml is saved to its own cache key (using the table id as part of its cache key). Only a table that has been modified will have its cache cleared.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
no FF

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Locally tested and tested on staging

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
I added tests that only when a table is modified is it included in fixture xmls that's created for the restore. Tests exists that checks fixture xml is created as expected in various scenarios.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
no QA

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [X] The migrations in this code can be safely applied first independently of the code

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

This PR includes a migration that adds `last_modified` fields for `LookupTable`. If this PR/the migration is reverted, we have to ensure no other changes have been made that uses this field.

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change


[USH-5833]: https://dimagi.atlassian.net/browse/USH-5833?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Jtang-1 committed:
feats: adds last_modified field to track when the model changes for more selective syncing

Jtang-1 committed:
feat: track timestamp for modification of table that resulted from deleting a row or owner that belonged to that table

the last_modified field of LookupTable will represent changes to the table model or deletion of related row or owners.

Jtang-1 committed:
feat: create query for tables that have been modified

Jtang-1 committed:
feat: helper function to check if full sync is needed

full sync will require getting more than just fixtures that have recently been modified

Jtang-1 committed:
feat: use all tables to create xml only if fully syncing. Otherwise, only needs to include tables that have been modified since the last sync

Jtang-1 committed:
feat: never rely on cache

cache is only cleared when a lookup table is modified. If we continue caching, then in instances where no lookup table is modified,
the cache will either contain 1. the most recent list of modified tables or 2. all tables (if full sync is done).

Scenario 1: Upon full syncing, it will still only include the most recent list of modified tables
Scenario 2: Upon incremental syncing, it will continue including all tables

Jtang-1 committed:
feat: use last_modified on LookupTable to reflect deletion of related LookupTableRow or Owner

note that "delete" function does not get called for bulk deletion

Jtang-1 committed:
test: add last_modified to expected response

Jtang-1 commented:
@coderabbitai review

Jtang-1 committed:
test: checks that only fixtures that have been modified are included in the restore

esoergel commented on corehq/apps/fixtures/fixturegenerators.py:
```diff
@@ -148,7 +148,7 @@ def get_global_items(self, global_types, restore_state):
                                            FIXTURE_BUCKET,
                                            '',
                                            data_fn,
-                                           restore_state.overwrite_cache)
```
Can you talk more about this change?  Is this understanding roughly correct:

Before this PR, all lookup tables were synced in all restores (except those that skipped all fixtures).  This includes a set of "global fixtures," which are the same for all users on the domain.  Because of this fact, it was prudent to compute the XML for these global fixtures once, store it in blobdb, and serve to users directly from there, skipping the steps of pulling from SQL and serializing as XML.  In some situations, CommCare mobile would send an "overwrite_cache" signal, which would cause the global cache to be recomputed.  Do you know what causes this flag to be set?

Now after this PR, users will sometimes receive only a subset global fixtures (those which have changed since their last sync).  This breaks the monolithic caching of all global fixtures for the domain.  Is that all accurate?

Is this cache used at all then?  I see it'd still be used in [case restores](https://github.com/dimagi/commcare-hq/blob/bc19f920b830d608116a3b2cee55cf08d48e66eb/corehq/apps/ota/case_restore.py#L50-L50), but that's more of an esoteric feature probably used on a much smaller set of domains - if that's all it's for, I'd rather move the caching logic there so we can skip it for everyone who would otherwise just overwrite it on every request.

Better yet though, would it make sense to switch to caching global fixtures individually rather than all together?  Then individual large fixtures would still only need to be computed once per domain, and the decision of which fixtures to include could still vary per user.

Lastly, the terminology is a little confusing - is a "global type" a `LookupTable` instance, and a "global item" a `LookupTableRow` instance?

esoergel replied:
Yep that's right, it'd be a key per table.  Probably only necessary for large tables over some threshold of rows, but doing a db query to count rows might be more overhead than just caching it either way.  I'm not certain.

Jtang-1 replied:
Changes to cache by individual tables: [227e024](https://github.com/dimagi/commcare-hq/pull/36091/commits/227e0241191f733907d90a405f019d120df66907) and [9314ca3](https://github.com/dimagi/commcare-hq/pull/36091/commits/9314ca3d8bf2bbfe12d7b2ca8a3a57df020c19f3)

Jtang-1 commented on corehq/apps/fixtures/models.py:
```diff
@@ -27,6 +27,29 @@ def by_domain_tag(self, domain_name, tag):
     def domain_tag_exists(self, domain_name, tag):
         return self.filter(domain=domain_name, tag=tag).exists()
 
+    def get_tables_modified_since(self, domain, since_datetime):
```
`LookupTableRow` and `LookupTable` are indexed by domain so I would guess that isn't expensive? Also it should just be one query. `LookupTableRowOwner` has 3.7 million rows so that seems like it'll be expensive. Would indexing that by `domain` also greatly help performance?

esoergel replied:
I'd probably time it on prod, actually, since it depends on the amount of data in the DB.  But yeah adding an index on `domain` does sound like it'd make a lot of sense if the query does look slow.  I bet there are also other ways to structure it, but I don't think I've got good intuition on what would be fastest.

Jtang-1 replied:
From Daniel's suggestion, I consolidated `last_modified` to be defined on only `LookupTable` model but it also captures when its related row or row owner is created, updated, or deleted. This allowed me to simplify this nasty query [e688bb3](https://github.com/dimagi/commcare-hq/pull/36091/commits/e688bb3cae8d67b3b8055a16cf1e2acb5ca43ab3)

Jtang-1 committed:
feat: we expet a large number of owner row ids to process in chunk for query performance

Jtang-1 commented on corehq/apps/fixtures/fixturegenerators.py:
```diff
@@ -151,6 +151,12 @@ def get_items_by_type(data_type):
+    def _should_full_sync(self, last_sync, app_id):
+        app_has_changed = (last_sync and last_sync.build_id is not None
+                and app_id is not None
+                and app_id != last_sync.build_id)
```
Right—I was thinking that lookup tables are only included in the restore if they are used in an app the user has access to, but that’s incorrect. So, I don’t think we need to check whether an app has changed.

However, we do check for app changes when dealing with location fixtures. After reading the related [PR](https://github.com/dimagi/commcare-hq/pull/23359) and ticket, I’m not sure why that’s necessary either. Does that ring a bell for you?

esoergel replied:
> Set the "App Aware Location Fixture Format" to Both Flat and hierarchical

That's the cause, I believe.  We have a deprecated behavior that involves a different XML structure for the location fixture, which is enabled only for certain applications.  Toggling this behavior by modifying the app means we need to re-sync.  This solution does seem like it would trigger _far_ more location syncs than necessary, however, as changing that setting is likely extremely rare, and it only applies to projects with the feature flag enabled anyways.  I bet we could further optimize that significantly.

Jtang-1 replied:
removed check for new app build. [afe41ef](https://github.com/dimagi/commcare-hq/pull/36091/commits/afe41ef265aca135b5243ba669bc1ca99e21b3f5)

dannyroberts commented:
I have a comment about the concept rather than the code. I hope it's helpful because I think it's non-obvious. On the face of it, this is a straightforward strict improvement: given a particular size lookup table, every restore request will be either as fast as or faster than it was before. Hiding underneath that obvious benefit is a new non-obvious risk: whereas before it was painfully clear to a project when a lookup table was too big—every single sync would take forever—and thus when they had to make hard decisions to conserve on lookup table space, now every time a lookup table is updated (maybe by just adding one or two items), each update is slightly more disruptive than the last one, but only for a one refresh cycle. The response will shift from "syncs are slow and we need to do something to optimize them" to "Well that was weird. Anyway...", until one day the lookup tables are so large that restores don't successfully complete and the project is ground to a halt.

People do tend to sync all at once, especially first thing when they get to work and first thing when they come back from lunch, for example, which can exacerbate the all-calm / everything-all-at-once effect bimodal behavior.

None of this is to say we shouldn't do it. But I think it's worth thinking through what those new "everything-all-at-once" moments will look like and how to prepare people/projects for them and/or make them more manageable. I'm not very up to date on how caching works here, but from your note on caching in the PR description, it sounds like we currently don't cache lookup table XML payloads or any such pre-built artifact. In the case where, say 10,000 people wake up to a lookup table update and all sync at once, and assuming they're all getting the same version of the table, it would make a big difference whether we're regenerating payload concurrently 10,000 times or generating it at most a few times (due to race conditions among the first requests) and then reusing that work for the other 9,990 times.

Some things to think about.

millerdev commented:
> One caveat is that [auto_now](https://docs.djangoproject.com/en/5.1/ref/models/fields/#django.db.models.DateField.auto_now) does not update on `update()` or `bulk_update()`, as it only triggers on `save()`. However, since these models are not currently updated via `update()` or `bulk_update()`, this limitation is not a concern.

Are you sure that's true?
https://github.com/dimagi/commcare-hq/blob/991500d6d8082358eea6ed5875a4b30f952cb20f/corehq/apps/fixtures/views.py#L205-L208

Rather than adding a `last_modified` field to `LookupTableRow` and `LookupTableRowOwner`, what do you think of instead always ensuring that `table.save()` is called any time a row or row owner is created/updated/deleted? I think that would have an added bonus of reducing the expense of determining if a table has been modified.

Jtang-1 committed:
full sync is not needed just because the app is changed. Lookup tables are domain level defined, not app specific

Jtang-1 commented on corehq/apps/fixtures/upload/run_upload.py:
```diff
@@ -163,11 +163,34 @@ def bulk_delete(model_class, to_delete):
+            owner_row_ids = [owner.row_id for owner in owners.to_delete]
+            owner_table_ids = set(LookupTableRow.objects.filter(
+                id__in=owner_row_ids
+            ).values_list('table_id', flat=True).distinct())
```
Yes good point, that should be done in chunks [2688c5f](https://github.com/dimagi/commcare-hq/pull/36091/commits/2688c5f9f4e1b02e4107e2b61b7fc237ea2c012c)

millerdev commented:
> This would involve overriding the create.. etc. methods right? I considered this but wouldn't this be nonperformant with bulk operations done to LookupTableRow or LookupTableRowOwner? As in, each save to table would be done individually and not benefit from the optimizations involved with bulk operations.

My proposal was worded poorly, and that's not quite what I had in mind. When I said

> what do you think of instead always ensuring that `table.save()` is called any time a row or row owner is created/updated/deleted?

I should have phrased it as

What do you think of instead always ensuring that any time a change operation is done on lookup table rows or owners, also endure that the modified date is updated on the parent lookup table. This would be done **once per _table_ change operation**, probably at the same point where we currently invalidate lookup table caches, for example after a lookup table upload is complete.

esoergel commented:
@dannyroberts 

> but from your note on caching in the PR description, it sounds like we currently don't cache lookup table XML payloads or any such pre-built artifact

We do, actually.  The set of tables that are the same for all users on the domain are rendered to XML and stored in blobdb.  However, this functionality is being dropped in this PR, though I think it'll end up just being changed.  See [this comment thread](https://github.com/dimagi/commcare-hq/pull/36091#discussion_r2019085881) for more.

> The response will shift from "syncs are slow and we need to do something to optimize them" to "Well that was weird. Anyway...", until one day the lookup tables are so large that restores don't successfully complete and the project is ground to a halt.

I hadn't considered this perspective.  I don't have any great ideas for how to surface that risk more clearly.  On USH projects we have pretty good metrics to identify stuff like this - if a small number of syncs were especially slow, that would show up on the BHA dashboard in datadog, for instance.  However, almost no projects have that degree of access and technical capacity.

The usage patterns for lookup tables do help a little here though.  Lookup tables I assume grow in an obvious, explicit manner.  When a project updates a large lookup table, it will invalidate the whole table, meaning every user's next sync will be slower accordingly.  I do think this is an easier pattern to spot than issues with casedb size for instance, which mostly grow slowly over time, user by user (except for large imports).

Jtang-1 committed:
Merge branch 'master' into jt/lookup-table-smart-sync

Jtang-1 committed:
feat: builds fixture serialized XML indivdually so that it can be cached separately.

Uses the date type's id as part of the cache key. Caching data types separately lets us avoid
rebuilding the XML for fixtures that have not been recently modified. It also allows us to continue
using caching when doing full syncs.

Jtang-1 committed:
feat: use key when writing to cache since there is no longer just one domain-fixtures cache per domain

Jtang-1 committed:
feat: data types are cached individually instead of grouped together by a domain key. So need to specify the data type id to clear the cache for

Jtang-1 committed:
feat: data types are cached individually instead of grouped together by a domain key. So need to specify the data type id to clear the cache for

Jtang-1 committed:
combine

Jtang-1 committed:
feat: use last_modified on LookupTableModel to capture whenever a related row or owner is created,changed,or deleted

This allows us to do one update per table change and groups it with invalidating cache. It
also reduces the expense of determining if a table has been modified.

Jtang-1 committed:
only need to query for LookupTable model since its last_modified field encapsulates changes to its related rows or row owners

Jtang-1 commented:
Thank you for the suggestion @millerdev. That really works out nicer considering the move to caching fixtures individually now as well. It also solved that expensive query that had to perform subqueries to determine if rows or owners were recently modified.

Jtang-1 committed:
test: last modified was removed from LookupTableRow model

Jtang-1 committed:
bug: handle when data_type_ids is a UUID instance

Jtang-1 committed:
test: pass in date type id when creating fixture bucket cache key

Jtang-1 committed:
test: fix test

Jtang-1 committed:
bug: get the id before the items are deleted

Jtang-1 committed:
test: implementation no longer contains last_modified fields for LookupTableRow and LookupTableRowOwner so saving and deleting those instances no longer trigger updating those field values. We now only look at `last_modified` value for LookupTable

Jtang-1 committed:
test: add table id argument to clear cache

Jtang-1 committed:
test: remove wrong commas

millerdev commented on corehq/apps/fixtures/utils.py:
```diff
@@ -55,6 +57,10 @@ def get_index_schema_node(fixture_id, attrs_to_index):
 
-def clear_fixture_cache(domain):
+def clear_fixture_cache(domain, data_type_ids):
+    if isinstance(data_type_ids, str) or not hasattr(data_type_ids, '__iter__'):
+        data_type_ids = [data_type_ids]
```
Always pass in a sequence of ids rather than doing type checking here.
```suggestion
    assert not isinstance(data_type_ids, str), 'expected list or set, got str'
```

Jtang-1 replied:
[51dfb13](https://github.com/dimagi/commcare-hq/pull/36091/commits/51dfb13ab90bb440308f2bde6bf489d515355181)

millerdev commented on corehq/apps/fixtures/views.py:
```diff
@@ -100,8 +100,9 @@ def update_tables(request, domain, data_type_id=None):
+            id = data_type.id
             data_type.delete()
-            clear_fixture_cache(domain)
+            clear_fixture_cache(domain, id)
```
```suggestion
            clear_fixture_cache(domain, [id])
```

Jtang-1 replied:
[51dfb13](https://github.com/dimagi/commcare-hq/pull/36091/commits/51dfb13ab90bb440308f2bde6bf489d515355181)

millerdev commented on corehq/apps/fixtures/views.py:
```diff
@@ -147,7 +148,7 @@ def update_tables(request, domain, data_type_id=None):
                 data_type = _create_types(
                     fields_patches, domain, data_tag, is_global, description)
-        clear_fixture_cache(domain)
+        clear_fixture_cache(domain, data_type.id)
```
```suggestion
        clear_fixture_cache(domain, [data_type.id])
```

Jtang-1 replied:
[51dfb13](https://github.com/dimagi/commcare-hq/pull/36091/commits/51dfb13ab90bb440308f2bde6bf489d515355181)

millerdev commented on corehq/apps/fixtures/tests/test_fixture_data.py:
```diff
@@ -246,11 +259,49 @@ def test_cached_global_fixture_user_id(self):
+        fixture = call_fixture_generator(self.user.to_ota_restore_user(self.domain), self.sync_log)
+        self.assertEqual(len(fixture), 0)
+
+        clear_fixture_cache(self.domain, self.data_type.id)
```
```suggestion
        clear_fixture_cache(self.domain, [self.data_type.id])
```

Jtang-1 replied:
[51dfb13](https://github.com/dimagi/commcare-hq/pull/36091/commits/51dfb13ab90bb440308f2bde6bf489d515355181)

millerdev commented on corehq/apps/linked_domain/updates.py:
```diff
@@ -384,8 +384,7 @@ def update_fixture(domain_link, tag, is_pull=False, overwrite=False):
         LookupTableRow.objects.bulk_create(chunk)
-
-    clear_fixture_cache(domain_link.linked_domain)
+    clear_fixture_cache(domain_link.linked_domain, linked_data_type.id)
```
```suggestion
    clear_fixture_cache(domain_link.linked_domain, [linked_data_type.id])
```

Jtang-1 replied:
[51dfb13](https://github.com/dimagi/commcare-hq/pull/36091/commits/51dfb13ab90bb440308f2bde6bf489d515355181)

millerdev commented on corehq/apps/ota/tests/test_case_restore.py:
```diff
@@ -119,6 +117,7 @@ def test_lookup_table_in_restore(self):
                             is_global=True,
                             fields=[TypeField(name="wing")])
         table.save()
+        self.addCleanup(clear_fixture_cache, self.domain, table.id)
```
```suggestion
        self.addCleanup(clear_fixture_cache, self.domain, [table.id])
```

Jtang-1 replied:
[51dfb13](https://github.com/dimagi/commcare-hq/pull/36091/commits/51dfb13ab90bb440308f2bde6bf489d515355181)

millerdev commented on corehq/apps/fixtures/resources/v0_1.py:
```diff
@@ -179,7 +179,7 @@ def obj_delete(self, bundle, **kwargs):
 
         query.delete()
-        clear_fixture_cache(kwargs['domain'])
+        clear_fixture_cache(kwargs['domain'], kwargs['pk'])
```
```suggestion
        clear_fixture_cache(kwargs['domain'], [kwargs['pk']])
```

Jtang-1 replied:
[51dfb13](https://github.com/dimagi/commcare-hq/pull/36091/commits/51dfb13ab90bb440308f2bde6bf489d515355181)

millerdev commented on corehq/apps/fixtures/resources/v0_1.py:
```diff
@@ -361,8 +361,9 @@ def obj_delete(self, bundle, **kwargs):
+        table_id = row.table.id
         row.delete()
-        clear_fixture_cache(row.domain)
+        clear_fixture_cache(row.domain, table_id)
```
```suggestion
        clear_fixture_cache(row.domain, [table_id])
```

Jtang-1 replied:
[51dfb13](https://github.com/dimagi/commcare-hq/pull/36091/commits/51dfb13ab90bb440308f2bde6bf489d515355181)

millerdev commented on corehq/apps/fixtures/resources/v0_1.py:
```diff
@@ -382,7 +383,7 @@ def obj_create(self, bundle, request=None, **kwargs):
             bundle.obj.save()
         finally:
-            clear_fixture_cache(kwargs['domain'])
+            clear_fixture_cache(kwargs['domain'], data_type_id)
```
```suggestion
            clear_fixture_cache(kwargs['domain'], [data_type_id])
```

Jtang-1 replied:
[51dfb13](https://github.com/dimagi/commcare-hq/pull/36091/commits/51dfb13ab90bb440308f2bde6bf489d515355181)

millerdev commented on corehq/ex-submodules/casexml/apps/phone/utils.py:
```diff
@@ -33,54 +33,65 @@
     io = BytesIO()
     io.write(ITEMS_COMMENT_PREFIX)
-    io.write(six.text_type(len(items)).encode('utf-8'))
+    io.write(six.text_type(len(io_streams)).encode('utf-8'))
```
`six` can be removed. It's long obsolete.
```suggestion
    io.write(str(len(io_streams)).encode('utf-8'))
```

FYI, I did not review other changes in this file carefully.

Jtang-1 replied:
[14a224f](https://github.com/dimagi/commcare-hq/pull/36091/commits/14a224fba97017994d42ba9ea38c78b6503b6330)

Jtang-1 committed:
test: set up cleanup as early as possible in case of failure in set up

Jtang-1 committed:
avoid hitting the database to lead the related row.table object

Jtang-1 committed:
six is obsolete

Jtang-1 committed:
always pass list for simplicity

Jtang-1 committed:
bug: it was not capturing all modified tables if there were more than 1000 rows. `flush` is called every 1000 rows and the mutation sets of tables, rows, and owners are cleared.

This implementation uses the table object that is already in memory and collects the modified tables during processing before flush occurs.

Jtang-1 commented on corehq/apps/fixtures/tests/test_fixture_data.py:
```diff
@@ -75,7 +79,16 @@ def setUp(self):
+        )
+        self.sync_log.save()
+
+        self.addCleanup(get_blob_db().delete, key=FIXTURE_BUCKET(self.data_type.id) + '/' + self.domain)
```
good point [20e2d78](https://github.com/dimagi/commcare-hq/pull/36091/commits/20e2d78c8820d5ac535039395da4beb16d04035d)

Jtang-1 commented on corehq/apps/fixtures/resources/v0_1.py:
```diff
@@ -361,8 +361,9 @@ def obj_delete(self, bundle, **kwargs):
             row = LookupTableRow.objects.get(id=kwargs['pk'])
         except LookupTableRow.DoesNotExist:
             raise NotFound('Lookup table item not found')
+        table_id = row.table.id
```
[30d3d43](https://github.com/dimagi/commcare-hq/pull/36091/commits/30d3d430d19f9d5a490381eceec1d1f2a29ab2ca)

Jtang-1 commented on corehq/apps/fixtures/resources/v0_1.py:
```diff
@@ -402,7 +403,7 @@ def obj_update(self, bundle, **kwargs):
                 bundle.obj.save()
             finally:
-                clear_fixture_cache(bundle.obj.domain)
+                clear_fixture_cache(bundle.obj.domain, bundle.obj.table.id)
```
[30d3d43](https://github.com/dimagi/commcare-hq/pull/36091/commits/30d3d430d19f9d5a490381eceec1d1f2a29ab2ca)

Jtang-1 commented on corehq/apps/fixtures/upload/run_upload.py:
```diff
@@ -104,10 +123,11 @@ def process_row(row, new_row):
             deleted_key=attrgetter("tag"),
         )
 
+        ids_of_modified_tables = get_ids_of_modified_tables(tables, rows, owners)
```
ah thank you for catching this mistake and for taking the time to make the code change diff. That is much better than the query I was doing. Also added tests in this commit [50bf7c1](https://github.com/dimagi/commcare-hq/pull/36091/commits/50bf7c1d430cbf170711121f296dbf15ca68ba83)

Jtang-1 commented on corehq/apps/fixtures/fixturegenerators.py:
```diff
@@ -136,20 +146,25 @@ def __call__(self, restore_state):
+        return [
+            (FIXTURE_BUCKET(global_type_id),
+             partial(self._get_single_global_item, global_type_id, global_type, domain))
+            for global_type_id, global_type in sorted(global_types.items(), key=lambda item: item[1].tag)
```
I carried it over from [here](https://github.com/dimagi/commcare-hq/blob/227e0241191f733907d90a405f019d120df66907/corehq/apps/fixtures/fixturegenerators.py#L185) so the the behavior is unchanged.

 It looks like sorting by tag was originally introduced [here](https://github.com/dimagi/commcare-hq/commit/a0788a1d5feda94e89782f95cc95cc21b88fd79a) just for consistency when generating the xml

millerdev commented on corehq/apps/fixtures/tests/test_upload.py:
```diff
@@ -684,6 +684,26 @@ def error_tx():
+
+        with patch.object(mod, "clear_fixture_cache") as mock_clear:
+            self.upload([(apple_id, 'N', 'apple'), (None, 'N', 'banana')])
+            self.assertEqual(len(mock_clear.call_args[0][1]), 1)
+            self.assertIn(table.id, mock_clear.call_args[0][1])
```
nit: I think reducing this to a single assertion will make test failures easier to debug if these assertions ever fail. This same idea can be applied to each of the new tests in this module.
```suggestion
            self.assertEqual(mock_clear.call_args[0][1], [table.id])
```

millerdev reviewed (commented):
LGTM for the parts that I reviewed and commented on.

Jtang-1 committed:
refactor: move conditional check into query filter

Jtang-1 committed:
refactor: make get_or_cache_global_fixture responsible for only one global fixture at a time

Jtang-1 committed:
refactor: DRY getting global items

Jtang-1 committed:
style: uppercase should only be used for global variables

Jtang-1 committed:
style: improve assertion message

Jtang-1 committed:
test: reduce to single assertion for easier debugging

Jtang-1 commented on corehq/apps/fixtures/fixturegenerators.py:
```diff
@@ -98,8 +98,12 @@ def get_global_items_by_domain(domain, case_id):
         if data_type.is_global:
```
[76e8fcc](https://github.com/dimagi/commcare-hq/pull/36091/commits/76e8fcc8d98a0d107e4b7aca74b3f52189b8e363)

Jtang-1 commented on corehq/apps/fixtures/fixturegenerators.py:
```diff
@@ -98,8 +98,12 @@ def get_global_items_by_domain(domain, case_id):
+        return get_or_cache_global_fixture(domain,
+                                           case_id,
+                                           cache_bucket_prefix_data_fn_pairs,
+                                           '')
```
good point, I was keeping it around in its more flexible form because it felt like a shame to delete code that could be useful... but I guess YAGNI
[4c0c337](https://github.com/dimagi/commcare-hq/pull/36091/commits/4c0c337d419eb30357f66a004624248cdb58986d)

Jtang-1 committed:
fix typo

Jtang-1 commented on corehq/apps/fixtures/utils.py:
```diff
@@ -58,8 +59,7 @@ def get_index_schema_node(fixture_id, attrs_to_index):
 def clear_fixture_cache(domain, data_type_ids):
-    if isinstance(data_type_ids, str) or not hasattr(data_type_ids, '__iter__'):
-        data_type_ids = [data_type_ids]
+    assert not isinstance(data_type_ids, (str, UUID)), 'expected list or set, got str or UUID'
```
Decided to assert for both just to be more explicit [e087116](https://github.com/dimagi/commcare-hq/pull/36091/commits/e087116f187f0b924e912fe063329b5288ebabd5)

Jtang-1 commented on corehq/apps/fixtures/fixturegenerators.py:
```diff
@@ -98,8 +98,12 @@ def get_global_items_by_domain(domain, case_id):
     if global_types:
-        data_fn = partial(ItemListsProvider()._get_global_items, global_types, domain)
-        return get_or_cache_global_fixture(domain, case_id, FIXTURE_BUCKET, '', data_fn)
+        cache_bucket_prefix_data_fn_pairs = ItemListsProvider()._get_fixture_cache_bucket_to_generator_pair(
```
yaa.... I'm open to any naming suggestions even aside from this specific one.

Jtang-1 committed:
test: expected set, not list

esoergel commented on corehq/ex-submodules/casexml/apps/phone/utils.py:
```diff
@@ -31,67 +30,24 @@
+    data = io.read()
+    global_id = GLOBAL_USER_ID.encode('utf-8')
+    b_user_id = user_id.encode('utf-8')
+    return [data.replace(global_id, b_user_id)] if data else []
```
Does this still need to write to `io` only to `read()` it immediately?  Previously I think it did that so it could save the `io` stream as a monolithic global cache.  Now that the `fixture_io_streams` going into this are being cached individually instead, I think this can be simplified.

esoergel replied:
Also another thing I'm thinking about here is performance - fixtures sometimes get so big that restores fail, so we should be sure not to accidentally lower the threshold at which that happens.  Here I see the full fixture element is being evaluated and held in memory at once, rather than being streamed through iteratively.  It looks like it did the same thing before, so I think this should be fine, just wanted to make sure it's something you're thinking about as well.

Jtang-1 replied:
[f8a4ebf](https://github.com/dimagi/commcare-hq/pull/36091/commits/f8a4ebf33f01ee694d765be9e3db5d1487cfb2dd)
I considered doing direct byte concatenation with `+=` but apparently it's not as performant since it creates a new bytes object on each addition and results in repeated memory allocations and data copying. `bytearray` allows for in-place extension so is more performant. This timing script suggests that `BytesIO` and `bytearray` have around the same performance so this change should be fine.

BytesIO:
  Max time: 0.0137 seconds
  Min time: 0.0021 seconds
  Mean time: 0.0042 seconds

bytearray:
  Max time: 0.0101 seconds
  Min time: 0.0033 seconds
  Mean time: 0.0045 seconds

Byte concatenation:
  Max time: 0.0782 seconds
  Min time: 0.0354 seconds
  Mean time: 0.0473 seconds
```
import time
from io import BytesIO
import statistics

# Create simulated large byte streams
NUM_STREAMS = 100
BYTES_PER_STREAM = 1024 * 100  # 100 KB per stream

fixture_io_streams = [
    BytesIO(b'x' * BYTES_PER_STREAM) for _ in range(NUM_STREAMS)
]

def using_bytesio(streams):
    io = BytesIO()
    for stream in streams:
        stream.seek(0)
        io.write(stream.read())
    return io.getvalue()

def using_concat(streams):
    data = b''
    for stream in streams:
        stream.seek(0)
        data += stream.read()
    return data

def using_bytearray(streams):
    data = bytearray()
    for stream in streams:
        stream.seek(0)
        data.extend(stream.read())
    return bytes(data)

def time_function(fn, label, num_runs=25):
    times = []
    for _ in range(num_runs):
        start = time.perf_counter()
        result = fn(fixture_io_streams)
        duration = time.perf_counter() - start
        times.append(duration)
    
    max_time = max(times)
    min_time = min(times)
    mean_time = statistics.mean(times)
    
    print(f"{label}:")
    print(f"  Max time: {max_time:.4f} seconds")
    print(f"  Min time: {min_time:.4f} seconds")
    print(f"  Mean time: {mean_time:.4f} seconds")

time_function(using_bytesio, "BytesIO")
time_function(using_concat, "Byte concatenation")
time_function(using_bytearray, "bytearray")
```

Jtang-1 committed:
refactor: avoid writing just to read immediately

timing tests suggests that bytearray is as performant as BytesIO when doing byte aggregation

Jtang-1 committed:
refactor: add domain direclty when generating key

Jtang-1 commented on corehq/apps/fixtures/tests/test_fixture_data.py:
```diff
@@ -44,6 +48,7 @@ def setUp(self):
             item_attributes=[],
         )
         self.data_type.save()
+        self.addCleanup(get_blob_db().delete, key=fixture_bucket(self.data_type.id) + '/' + self.domain)
```
i like that [8872bf6](https://github.com/dimagi/commcare-hq/pull/36091/commits/8872bf6ee3fb198ba7b698c976c22f3ae2971f76)

esoergel commented on corehq/ex-submodules/casexml/apps/phone/utils.py:
```diff
@@ -33,91 +29,53 @@
-                data = get_cached_fixture_items(domain, cache_bucket_prefix)
-                if data is not None:
-                    return [data]
+def combine_io_streams(fixture_io_streams, user_id):
```
Caveat: I think it's fine to punt on this, but wanted to mention as it's something I noticed.

I don't think this actually needs to combine the io streams at all anymore.  The old code did this:
Generate some number of fixtures, concat them all together, and store a comment in the XML indicating how many fixtures it contained.
Then this cached value was fetched and returned as one bytestring in a list from `ItemListsProvider.__call__`.
That gets called [here](https://github.com/dimagi/commcare-hq/blob/1ac764c14bf94296df02138e001295bb8af14ce9/corehq/ex-submodules/casexml/apps/phone/fixtures.py#L61-L62), which is called by something that just calls [this append method](https://github.com/dimagi/commcare-hq/blob/1ac764c14bf94296df02138e001295bb8af14ce9/corehq/ex-submodules/casexml/apps/phone/restore.py#L127-L134) on each element in the list.
That `append` method parses the comment to get the number of elements stored in the bytestring, then discards the comment and appends the bytestring.

I think instead, we could:
Make `ItemListsProvider.__call__` return each component of the large bytestring as its own element.  They would still get appended together in the same order (through multiple calls to `append`), so it should have no functional difference in how the restore turns out.  That would also let us drop the comment containing the count, the thing parsing and removing the comment, and the code combining the streams here.

millerdev replied:
>  That would also let us drop the comment containing the count, the thing parsing and removing the comment, and the code combining the streams here.

+1 This seems like a significant improvement.

Jtang-1 replied:
Ah thanks for pointing that out, i didn't consider what the original purpose of the comment was. Your suggestion makes sense to do, I made a ticket for it https://dimagi.atlassian.net/browse/USH-5894
